source/building-your-own-javascript-components/index.html.md.erb

```diff
@@ -127,13 +127,13 @@ class LinkEnhancement extends Component {
 
 As `govuk-frontend` does not provide type definitions in its package, TypeScript might show the following error message: 
 
-<q>Property ‘$root’ does not exist on type &lt;NAME_OF_YOUR_CLASS&gt;.</q>
+<q>Property '$root' does not exist on type &lt;NAME_OF_YOUR_CLASS&gt;.</q>
 
-You can work around this issue by defining the ‘$root’ property in your component class yourself.
+You can work around this issue by defining the `$root` property in your component class yourself.
 
 ```js
 class MyComponent extends Component {
```

Comment:
@romaricpascal How crucial is it that the `@template` (or TypeScript generic) type is explained?

```suggestion
class MyComponent extends Component<HTMLSelectElement> {
```

Reply:
I'll double check with the team, but my personal view on it is that until our team starts maintaining types (either within or outside the package), it's not up to our documentation to provide explanations around how to type things.

This bit was more about unblocking people as we realised that by moving the assignment of `this.$root` in `Component`, TypeScript wasn't aware of its existence in the methods of classes that extend `Component` because we don't ship types. As this is was introduced by this feature, we thought it'd be useful to explain a workaround and not fully block people using TypeScript.

The workaround we describe is about to change, as well, as we've since discovered that a common way to go when a library does not provide types seems to be [to use a `declare module` statement in a `.d.ts` file](https://github.com/alphagov/govuk-design-system/pull/4207). It types everything as `any` for `govuk-frontend`, but that's the same as what's happening for the exports of `govuk-frontend` in the absence of type definitions.

We're still on the fence about maintaining types, hence the last statement in the section to try and get more input on who consumes GOV.UK Frontend with TypeScript and how big a need it would be for us to ship these types.

Reply:
Ah fair enough, it was a surprise to see "with TypeScript" guidance so wanted to make sure

Let's keep hoping type-only [field declarations](https://www.typescriptlang.org/docs/handbook/2/classes.html#type-only-field-declarations) come to JSDoc as all of this would work 😣

```ts
class Example {
  declare $root: HTMLElement
}

class Custom extends Example {
  declare $root: HTMLButtonElement
}
```

>The workaround we describe is about to change, as well, as we've since discovered that a common way to go when a library does not provide types seems to be https://github.com/alphagov/govuk-design-system/pull/4207. It types everything as any for govuk-frontend, but that's the same as what's happening for the exports of govuk-frontend in the absence of type definitions.

Yeah, sorry it wasn't my place to say

It's similar to what others have done though:
https://github.com/hmcts/expressjs-template/blob/master/src/main/types/govuk-frontend.d.ts

A few more lines and you can write your own types:
https://github.com/dfe-analytical-services/explore-education-statistics/blob/dev/src/explore-education-statistics-common/src/types/govuk-frontend.d.ts

Reply:
> Yeah, sorry it wasn't my place to say

No worries, you couldn't be aware we were looking to change it 😊 

Cheers for the links to what services do for defining GOV.UK Frontend's types, as well.